doc/compiler/AOTClassChains.md

```diff
@@ -0,0 +1,153 @@
+JVM instance, the AOT infrastructure processes the Relocation Records; 
+it finds a candidate `J9Class` pointer in the current JVM and walks its 
+superclasses and interfaces, checking each one's romclass against the one 
+in the Class Chain stored in the SCC. Any mismatch implies the classes 
+might be different, in which case the AOT load fails. In the case of 
+inlined methods however, the AOT load doesn't necessarily fail, rather, 
+the AOT infrastructure disables a particular inlined call site.
+
+## Finding a candidate J9Class
+Finding a candidate J9Class with which to perform the validation is 
+non-trivial. Java uses Class Loaders to load classes. Two class loaders 
+can load the same "C extends B extends A" class twice, and this will 
+result in two different `J9Class` pointers; the difference comes from 
+the fact the two class loaders are different. Frameworks such as OSGI 
+uses different class loaders for different modules, and Java 
+applications can use several OSGI modules. Even if there weren't 
+multiple instances of a particular class in the current JVM instance, 
+having to go through every possible class loader to check if it already 
+has that particular class loaded would be extremely slow. If there were 
+multiple instances, then the AOT infrastructure would additionally need 
+to validate every single one. All this would effectively defeat the 
+purpose of AOT code, which is to facilitate fast application startup.
+
+Finding a particular `J9Class` in the current JVM instance requires 
+a Class Loader and a Class Name. The Class Name can be obtained from 
+the top romclass in the class chain. However, finding the Class Loader 
+is not as simple since they are just Java Objects created by a Java 
+application. OpenJ9 solves this problem by building has tables that 
+map a) Class Loaders to the first Class that loader loaded, and b) the 
+first class that loader loaded to the Class Loader Object. These tables
+are not persisted, it is built up via the Class Load Hook in the current 
+JVM instance. When referencing a particular `J9Class` in a Relocation 
+Record, the JIT stores 1) the offset to the class chain for the `J9Class` 
+itself, as well as 2) the offset to the class chain for the first 
+loader that loaded the `J9Class`.
+
+An an AOT load, the AOT infrastructure takes 1) and adds to the 
+current address of the SCC in the current JVM instance to find the class 
+chain to be validated. This class chain provides the name of the class. 
+The AOT infrastructure then uses 2) to get the Class Loader Object in 
+the current JVM instance, which will likely give the correct Class Loader. 
+Using the Class Name, the Class Loader then provides a `J9Class` pointer 
+which is used to perform the validation.
+
+## Subtleties
+For classes that are directly referenced in a method's code, for example 
+class `D` in `C.foo()` if `D.bar()` was a static call, there is a 
```

Comment:
the wording to identify the class 'C' in a static call to D.bar() made by C.foo() here is a bit confusing, you may want to expand a bit to make it more clear